Use memory management hardware to handle faults, measure user strings

tyrantlizard77 opened the pull request:
Some forthcoming system calls have user mode pass in strings which could be potentially garbage and generate an exception.

Trap memory management hardware exceptions during these operations and gracefully return an error instead.

Commit:
x86: add z_arch_user_string_nlen

Uses fixup infrastructure to safely abort if we get a page
fault while measuring a string passed in from user mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: add functions for copying to/from user

We now have functions for handling all the details of copying
data to/from user mode, including C strings and copying data
into resource pool allocations.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: add z_arch_user_string_nlen

Uses fixup infrastructure to safely abort if we get an MPU
fault when examining a string passed in from user mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arc: add z_arch_user_string_nlen

Safely measure the length of a potentially dubious string.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: add syscalls test case

Test that we can define our own system calls in application code
and that fault handling works properly.

Additional tests for base system call infrastructure, outside of
specific system calls, go here.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add z_arch_user_string_nlen prototype

This is used to measure the length of potentially unsafe
strings.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on arch/arc/core/fault.c:
```diff
@@ -26,15 +26,24 @@
+	/* TODO Improve this so that this logic only runs on MPU memory access
+	 * faults
+	 */
+	if (_current->fixup) {
```
Are we sure that the fault handling in all of the userspace-supported platforms are not nested?

lpereira reviewed (dismissed):
I like the approach, but I would have to take a look with care every place where the `fixup` pointer is set to ensure that this can't be abused as a generic indirect jump gadget.

pizi-nordic commented on kernel/userspace.c:
```diff
@@ -475,6 +475,114 @@ void _k_object_uninit(void *object)
+	char *ret = NULL;
+
+	key = irq_lock();
+	actual_len = z_user_string_nlen(src, maxlen, &err);
```
Could we just use retval < 0 as an error indication?

andrewboie replied:
I wanted to keep this the same as strnlen which returns a size_t.
Is there some advantage to this I am missing?

pizi-nordic replied:
Only simpler use and easier error dispatching (but this is my personal opinion).
```
ssize_t size = z_user_string_nlen(src, maxlen);
if (size < 0) {
      return -EFAULT;
}
```

pizi-nordic commented on kernel/include/syscall_handler.h:
```diff
@@ -125,6 +125,116 @@ extern void _thread_perms_all_clear(struct k_thread *thread);
+ * string, you will need to still do a Z_SYSCALL_MEMORY_READ()
+ * with the obtained size value to guarantee this.
+ *
+ * @param src String to measure size of
```
We should mention that this function must be called with interrupts disabled, due to interaction with fault handler.

andrewboie replied:
I don't believe this is actually necessary.
The fixup pointer is stored on a per-thread basis.
We can context switch out unless I'm missing something.

pizi-nordic replied:
What about situation when z_user_string_nlen() is being interrupted by a higher priority interrupt which triggers a fault? By my understanding, we will get a false positive here - instead of crashing kernel, we will assume that fault was caused by user string checking.

andrewboie replied:
Ah I see now.
Will address in next iteration of this PR, I have an idea on how to do this without requiring interrupt locking.

andrewboie replied:
New approach taken, interrupt locking should no longer be necessary.

pizi-nordic replied:
Now it looks better.

andrewboie commented:
Updated. The thread-specific fixup member is gone. Instead we build up a table of instruction pointer ranges with associated fixup handlers and use that instead. The fixup is only jumped to if the fault occurs within a very narrow range of instructions and nowhere else. 

Test cases augmented.

Just need some more testing on  ARC, having some trouble with the nsim target.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -640,11 +662,24 @@ static u32_t _FaultDump(const NANO_ESF *esf, int fault)
+
+		if (esf->pc >= start && esf->pc < end) {
+			esf->pc = (u32_t)(exceptions[i].fixup);
+			return;
```
As I understand, this code block will make `_Fault()` to by-pass the standard fault handling process handling. So we need to **manually** perform some actions that are now incorporated into the relative fault handlers (e.g. clear fault status register sticky bits, reset the fault address register etc.)

andrewboie replied:
> by-pass the standard fault handling process handling

Where does this take place? I cannot find it. All _Fault seems to do is call various code to dump out information and then call SysFatalErrorHandler() which then returns.

ioannisg replied:
In line 678 we seem to return from _Fault(). So, for example, if this was a BusFault, we are not going to clear the Sticky bits.

ioannisg replied:
The fault handlers do not only dump information, but also clear fault status.

ioannisg replied:
I am suggesting to refactor and clean up this code in Fault.c,so fault dumping and fault status clearing are somewhat separated....

andrewboie replied:
Well, this is a bad situation because in this scenario I really don't want anything printed.
Not only that, but this code doesn't even run if CONFIG_FAULT_DUMP isn't set. And I have to propagate back to the main _Fault() somehow that we need to return without calling SysFatalErrorHandler.
Do you have any advice other than rewriting fault.c for ARM?

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -640,11 +662,24 @@ static u32_t _FaultDump(const NANO_ESF *esf, int fault)
 	u32_t reason = _NANO_ERR_HW_EXCEPTION;
 	int fault = SCB->ICSR & SCB_ICSR_VECTACTIVE_Msk;
 
+#ifdef CONFIG_USERSPACE
```
As far as I understand, only memory protection faults could fall into this category of errors. So, MemMg for ARM and BusFaults for NXP, so I wonder if we could move the code inside the respective handlers.

This comment is related to my other comment below.

andrewboie replied:
I think this would require significantly restructuring the code in fault.c.

We don't really have handlers in here, just a bunch of stuff which prints out information to the console. There's no leeway for policy decisions to be made in the way it is structured now. All the "handlers" like _MpuFault are called by FAULT_DUMP, which only calls _FaultDump() if CONFIG_FAULT_DUMP > 0 and all the return values are ignored. It's a mess. 

Can we file a bug on this?

Commit:
arch: introduce exc_handle.h

Defines some common macros and data structures for declaring
text section ranges with fixup handler addresses if an
exception occurs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Moved some declarations to a header. Fixed some typos. Declared the exception handler arrays as const to put them in ROM.

Filed #8950 to figure out what to do about the ARM fault handling code buried in the fault dumping infrastructure, this still needs to be addressed.

Currently debugging some issues on ARC, I'm getting a hang when I test this on nsim_em.

andyross reviewed (commented):
Do you really need the arch assembly to implement the whole strlen loop?  Seems like a mostly-needless burden on the arch developer.  You could stuff the "fault_start" and "fault_end" symbols into simple C code with inline assembly and then just do the loop with whatever.

If there's concern that it be super-optimized, you could allow for an Z_ARCH_STRLEN() macro or whatever that inserts an optimized strlen sequence into the current C code, but default architectures would just pick up the "while(*s[len++]);" implementation.

andrewboie commented on arch/arm/core/fault.c:
```diff
@@ -619,6 +619,28 @@ static u32_t _FaultDump(const NANO_ESF *esf, int fault)
 }
 #endif /* FAULT_DUMP == 2 */
 
+#ifdef CONFIG_USERSPACE
```
I moved most of this to a header.

vonhust commented:
Generally OK for ARC,  my concerns are
1. avoid the abuse of this mechnism to weak the reliability and security of system

2. scalability. strnlen is one case,  i hope this mechnism can cover other possible cases to avoid repeat of work

andrewboie commented:
> Generally OK for ARC, my concerns are

Can you be more specific with your concerns?

> avoid the abuse of this mechnism to weak the reliability and security of system

Linux does essentially the same thing: https://github.com/torvalds/linux/blob/master/Documentation/x86/exception-tables.txt

> scalability. strnlen is one case, i hope this mechnism can cover other possible cases to avoid repeat of work

How is this not scalable?
Just add more items to the exception array, as long as their implementations have start/end/fixup addresses this will work.

andrewboie commented:
@ioannisg can you re-visit your review

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -164,13 +165,34 @@ static void _FaultShow(const NANO_ESF *esf, int fault)
  * @return 1 if error is recoverable, otherwise return 0.
  */
-static int _MpuFaultIsRecoverable(const NANO_ESF *esf)
+static int _MemoryFaultIsRecoverable(NANO_ESF *esf)
```
I am fine with changing MPU to Memory, if this function is to be used in a more generic way.

ioannisg replied:
@andrewboie , in the  amended version of this PR, this function is also used in ARMv6-M, so it needs to be  implemented outside the #ifdef ARMV7-M guard.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -268,8 +290,13 @@ static u32_t _MpuFault(const NANO_ESF *esf, int fromHardFault)
-static void _BusFault(const NANO_ESF *esf, int fromHardFault)
+static int _BusFault(NANO_ESF *esf, int fromHardFault)
 {
+	/* Assess whether system shall ignore/recover from this MPU fault. */
```
I suggest to put this inside #ifdef NXP_MPU,  as this is only relevant for NXP SOCs that generate BusFaults for memory access violations.

andrewboie replied:
This isn't true.
For example, in the provided test case, user mode passes a nonsense string address of 0xFFFFFFF0. This generates a bus fault.

ioannisg replied:
Ah, I see this possibility now, thanks for pointing it out, @andrewboie

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -268,8 +290,13 @@ static u32_t _MpuFault(const NANO_ESF *esf, int fromHardFault)
 {
+	/* Assess whether system shall ignore/recover from this MPU fault. */
+	if (_MemoryFaultIsRecoverable(esf)) {
+		return _NANO_ERR_RECOVERABLE;
```
The immediate `return `here is a bit problematic, because we need to clear the sticky bits, see line 386-371. I see two "good " options:
1. Copy here the code for sticky bits clearing (note that it is only relevant for ARMv8-M), or
2.  Create a local variable `reason `to hold the `_NANO_ERR_RECOVERABLE`, and jump with `goto `to the  sticky bit clearing and then `return reason`.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -342,6 +369,8 @@ static void _BusFault(const NANO_ESF *esf, int fromHardFault)
 	SCB->CFSR |= SCB_CFSR_BUSFAULTSR_Msk;
 #endif /* CONFIG_ARMV8_M_MAINLINE */
+
+	return _NANO_ERR_HW_EXCEPTION;
```
If you implement my second suggestion, here you could `return reason`.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -478,7 +507,7 @@ static u32_t _HardFault(const NANO_ESF *esf)
 			reason = _MpuFault(esf, 1);
 		} else if (SCB_BFSR) {
-			_BusFault(esf, 1);
+			reason = _BusFault(esf, 1);
```
good

ioannisg reviewed (changes requested):
I left some requests for changes.  But also: note that we still do not handle this properly for ARMv6-M /ARMv8_M_Base, where HardFault is the only fault exception.

I wonder whether we could, for now, support the whole thing  _only_ for ARMv7-M/ARMv8-M_Main

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -463,7 +492,7 @@ static void _DebugMonitor(const NANO_ESF *esf)
  * @return error code to identify the fatal error reason
  */
-static u32_t _HardFault(const NANO_ESF *esf)
+static u32_t _HardFault(NANO_ESF *esf)
```
@andrewboie, @agross-linaro , this is a comment related to adding support for recoverable memory-protection faults for ARMv6-M/ARMv8-M Baseline:  is it _as easy_ as just calling the MemoryFaultIsRecoverable(.) function in line 501 (i.e. for ARMv6 only) and do similar things as @andrewboie does in Bus and MPU fault?

ioannisg replied:
We will also need to move the print-out of "*** HARD FAULT ***" below the call of _MemoryFaultIsRecoverable, to avoid any fault dumping in this case...

andrewboie replied:
I think at this point I've let go of my desire to not have any printouts for recoverable faults, it just makes the code excessively complicated.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -463,13 +489,16 @@ static void _DebugMonitor(const NANO_ESF *esf)
 	PR_FAULT_INFO("***** HARD FAULT *****\n");
 
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
+	if (_MemoryFaultIsRecoverable(esf)) {
```
good

ioannisg reviewed (approved):
Going to approve this, however, I think we should not abandon the ambition of preventing fault dumping in case of errors that are supposed to be recoverable. We can merge this PR for 1.13, and do the refinement  under a future task, IMO.

ioannisg commented on arch/arm/core/fault.c:
```diff
@@ -164,13 +165,34 @@ static void _FaultShow(const NANO_ESF *esf, int fault)
 /* HardFault is used for all fault conditions on ARMv6-M. */
 #elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
 
+#ifdef CONFIG_USERSPACE
```
Should this also be outside the #ifdef ARMv7-M guard?

ioannisg reviewed (commented):
Added a few comments, after checking the sanity check errors.

AdithyaBaglody reviewed (approved):
LGTM, except for the error codes
